### PR TITLE
fix(eslint-plugin): [member-ordering] get accessor member name & take into account `abstract` and decorator

### LINKED ISSUE #9765: Bug: [@typescript-eslint/member-ordering] no error raised when accessors not sorted by alphabetical order
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.5.4&fileType=.ts&code=KYDwDg9gTgLgBAE2AMwIYFcA28DGnUDOBcA3gFBxxjoBGmAljnKjjsEdMwFxwExT0AdgHM4AXjgByVJLIUqtBkxZsOUODh58BI8VJyz5AeiNwAygAsIWBHCip6BYM0FxgUKJwJC2cAAY0fnCOcDQo0M5%2BOEFCcDAWzqiYYBaoYTCMSXDQSFAANKHo8AgQ7HCCEDDy1HSMzKzsBJw0WvxCohKSNIaURgBUcACSyHAAntbBggBu7vABQaiCtn7AABQAlH4F8c7IEJiYEADu7W4enCcHoc72jsAIXPKUkgCywAC2YeqCWJi8VjZrohgHhUFB7tc9uDmFcakpJnxFr53sB4hBbEhkEJ6BkIIICAA6OAARgAbFwAMxwAC0JPJFIALLJKJRVkdUvAUYtiPFUHNAsFiOCcBBhIJ6AAvCGEZjlPHUoSIwTI1FWBAFJpxBJ2LCJMBgBhlI7AA55J6FeA4mGarn4rV8rU3XVaoQAax5HP8ApCglQKOWPwOfnW8j6Rmqijqa3WPCmEHotnILLsqPQUEE8gAvnJeqYAEoOJwuM6edTeZWRBAxYhhKGRYAxVw7GEpNKozJ-HLuDUQR06zB6g30I0mzAR2pMBAbWPxxPm8EwNMZyjZzNAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6AW0XYCNFpaA9tAAmfZgHN0YANrhsOPtCGQANHPlZ52SKIBmAQwT4pmrds48%2BAFRIopss2cj6yZFMmVrHWyJ3wALAWFIdUcAXS9vKCFRaClIJn18WGh9eBDvAF9QiGz5MLlszKA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkAJutKgIabMHEVh6YCu6AGnA16jFmw5EAss0Rce-ITSipkAIy4BtYcvoJIAqOhJF0ADzyQdYALpKR8WHV4YukIk-SmLB6%2B9gAkkiq0Mh42ABuWJjIDPJ8gn6wvHgAIsiYbnTIJJb2tJiwsGkZbiSY0L66JADWyIgAMuoAwgAWjDXxin4kyRXoMnKUConVPMjQlsMJ%2BVB4rADm6FNGJABMAAxra1YQAL4ge0A&tokens=false

### Repro Code

```TypeScript
export default class {
  public accessor a: string = 'a'

  public accessor c: string = 'c'

  // Should raise an error since `b` is before `c` in the alphabetical order, but does not
  public accessor b: string = 'b'

  /* If you invert `b` and `e()`, the following error will be raised:
    'Member null should be declared before all public instance method definitions. 16:3 - 16:34'
    (what means that `b` is recognized as a non-instance method, so the rule applies well,
    but it also means that the rule thinks that `b` is named `null`)
  */
  public e(): void {
    return
  }

  // Raise an error since `d` is before `e` in the alphabetical order, so the rule applies well
  public d(): void {
    return
  }
}
```

### ESLint Config

```json
{
  "rules": {
    "@typescript-eslint/member-ordering": [
      "error",
      {
        "default": {
          "memberTypes": [
            "accessor",
            "method"
          ],
          "order": "natural"
        }
      }
    ]
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "declaration": true,
    "declarationMap": true,
    "lib": [
      "dom", "esnext"
    ],
    "module": "nodenext",
    "noImplicitOverride": true,
    "outDir": "dist",
    "rootDir": "src",
    "skipLibCheck": true,
    "sourceMap": true,
    "strict": true,
    "target": "es2022"
  }
}
```

### Expected Result

I expected the 7th line should report the error:
```
Member b should be declared before member c. 7:3 - 7:34
```

### Actual Result

There was no error on 7th line.

### Additional Info

If you invert `b` and `e()`, the following error will be raised:
```
Member null should be declared before all method definitions. 16:3 - 16:34
```
what means that `b` is recognized as a non-method, so the rule applies well, but it also means that the rule thinks that `b` is named `null`.

If you invert `accessor` and `method` in the eslint config:
```json
{
  "rules": {
    "@typescript-eslint/member-ordering": [
      "error",
      {
        "default": {
          "memberTypes": [
            "method",
            "accessor"
          ],
          "order": "natural"
        }
      }
    ]
  }
}
```

`a`, `b` and `c` are well recognized as `accessors` according to the new error messages:
```
Member e should be declared before all accessor definitions. 14:3 - 16:4
Member d should be declared before all accessor definitions. 19:3 - 21:4
```

**My assumption is that each `accessor` is treated as if it were named `null`.**

### PR DESCRIPTION
**Pull request by OlivierZal**
## PR Checklist

- [x] Addresses an existing open issue: fixes #9765
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

### CODE REVIEW

**Comment by OlivierZal** [comment]
Hi @Josh-Cena, I’m new to the typescript-eslint project. It seems I have a build issue, do you know how I can fix it?

**Comment by OlivierZal** [comment]
Hi @Josh-Cena, @JoshuaKGoldberg, @bradzacher,

How can I solve https://cloud.nx.app/runs/xRpPxqp5jU/task/visitor-keys%3Alint ?

**Comment by bradzacher** [comment]
image
It's a flake

**Comment by OlivierZal** [comment]
test1 non-abstract
test2 abstract

**Comment by OlivierZal** [comment]
Test with non-decorated accessor and decorated accessor:

test decorated

Test with non-abstract accessor and abstract accessor:

test abstract 2

**Comment by Josh-Cena** [comment]
Could you add _real_ test cases in the `tests` folder too?

**Comment by OlivierZal** [comment]
done @Josh-Cena

**Comment by OlivierZal** [comment]
@Josh-Cena, added test (and [fixed format](https://github.com/typescript-eslint/typescript-eslint/pull/9775/commits/d2f31472102f2922fee5ddc028aa78a5cec5d651) because latest commit broke `main`).

Now it's ready to merge.

**Review by JoshuaKGoldberg — APPROVED** [review]
Kid from the Brady Bunch saying "That's swell!"

**Comment by OlivierZal** [comment]
Thanks @JoshuaKGoldberg, is there any step I need to perform after approval is granted?

File: packages/eslint-plugin/src/rules/no-unnecessary-type-parameters.ts [hunk 1]
```diff
@@ -122,9 +122,7 @@ function isTypeParameterRepeatedInAST(
 
     // If the type parameter is being used as a type argument, then we
     // know the type parameter is being reused and can't be reported.
-    if (
-      reference.identifier.parent.type === AST_NODE_TYPES.TSTypeReference
-    ) {
+    if (reference.identifier.parent.type === AST_NODE_TYPES.TSTypeReference) {
```
Comment (Josh-Cena) [inline packages/eslint-plugin/src/rules/no-unnecessary-type-parameters.ts:125]: Hmmm, why is this changed?

**Review by Josh-Cena — APPROVED** [review]
Looks good to me :)

File: packages/eslint-plugin/src/rules/no-unnecessary-type-parameters.ts (hunk 1 above)
Comment (OlivierZal) [reply packages/eslint-plugin/src/rules/no-unnecessary-type-parameters.ts:125]: @Josh-Cena, explained [there](https://github.com/typescript-eslint/typescript-eslint/pull/9775#issuecomment-2283407308)

**Comment by Josh-Cena** [comment]
It seems wholly unnecessary to enforce up-to-date branches before merging in _most_ cases, FWIW.

**Comment by JoshuaKGoldberg** [comment]
> any step I need to perform after approval is granted

Just general excitement 😄

**Comment by JoshuaKGoldberg** [comment]
> any step I need to perform after approval is granted

Nothing on your end! Just being excited in general, I suppose. 😄 

Thanks for this! Merged & will be released in our weekly release later today.

**Pull request merged**